litestar/app.py

```diff
@@ -315,6 +314,9 @@ def __init__(
                 connections.
             experimental_features: An iterable of experimental features to enable
         """
+
+        self._openapi_factory = OpenAPIFactory(self)
```

Comment:
Is this actually needed here? If all it does is hold a reference to the app at this point, why not just initialise the attribute as `None` here (or don't set it at all and just declare it as a class attribute) and instantiate it on the first access to `openapi_schema`? That would allow to skip the "check if initialised" logic inside the factory

Reply:
Glad you brought it up.

> That would allow to skip the "check if initialised" logic inside the factory

This logic then gets moved outside of the factory, as we call `self._openapi_factory.add_route()` for each http route, which we'd have to guard with `if self._openapi_factory is not None:`.

I'm not really happy with the pattern here, but I don't think that delaying instantiation of the factory type really fixes much.

What I don't like about the way that we have this openapi stuff setup, is that there is special cased openapi schema logic dispersed across different methods of the `Litestar` object - it makes it a bit hard to reason about how it all works.

- In `Litestar.init()` we register the openapi controller 
- In `Litestar.register()` we only call `update_openapi_schema()` if `Litestar._openapi_schema` is not `None`.
- In `Litestar._openapi_schema` we raise if `Litestar.openapi_config` is `None` and if `Litestar._openapi_schema` is `None`, then call `Litestar.openapi_config.to_openapI_schema()` to create the schema, and then call `Litestar.update_openapi_schema()` to update it.
- In `Litestar.update_openapi_schema()` we raise an error if `Litestar._openapi_schema` is `None`, and also if `Litestar._openapi_schema.paths` is `None`. We also raise if there is no `Litestar.openapi_config`. I think at this point, these checks are to do with type narrowing within that method body, and not any functional requirement, but that's non-obvious when reading the code. We then go on to add route handlers to the schema, but only if the path for the route doesn't already exist (thinking out loud, but what if a handler for an existing path is added after the schema has been built?).
- After `Litestar.openapi_schema` has been accessed once, we then call `Litestar.update_openapi_schema()` every time that `Litestar.register()` is called.

It all just feels a bit unsettled to me.

Maybe a plugin would b a better solution, but we don't have a plugin that gets access to routes and handlers as they are added to the application.

So, say we expand plugins protocols to include a `ReceiveRoutePlugin` protocol that has a `receive_route()` method.

We could then build a plugin whose job it is to handle all of these gory details, and it woudn't necessarily need to even be a public plugin, we could just instantiate it and add it to the application's plugin registry if we receive an `OpenAPIConfig` object on initialization. The plugin could:

1. add the openapi controllers to the app route handlers
2. receive the routes as they are added to the application, determine the handlers on the route that should be included in the schema and keep a reference to them and their route.
3. handle deferred building of the openapi schema when it is first accessed. 
4. cache the openapi schema in app state instead of on the app object as currently occurs, removing a special case that doesn't really need to exist.
5. create dependency providers that provide the schema and its serialized versions for the openapi controllers with caching. This could then also handle invalidation of those cached objects better as FWICT that doesn't happen with the openapi controller as it stands, i.e., the json version of the schema is generated and cached and served, but if a route is dynamically added, that schema wouldn't be updated.
6. provide a callable in its namespace that retrieves the openapi_schema from an app instance which we could use wherever we currently access the schema from the app (in `app.openapi_schema` and by the CLI schema method).

We'd end up caching things on app state, using plugins to dynamically add the openapi controller, and have the schemas available via DI, which are all standard ways of doing these things in Litestar, and at the same time we'd remove a lot of special casing that occurs on the app object for the OpenAPI stuff.

Another use for expanding our protocols to receive routes would be the serialization plugins. If these things defined a `receive_route()` method (or maybe even a `receive_route_handler()` method) they'd be able to look at the handlers as they are added to the application, and inspect if they have had a DTO defined or not, and set a dto type for the handler at that point if they support the annotation types. This would save us from having to iterate over the serialization plugins twice for every handler to determine if a plugin should provide the DTO.

Reply:
Have a look at the latest commit to this branch to get a feel for it: https://github.com/litestar-org/litestar/pull/2805/commits/bbc1d1e059be74fec70ed96ddc0447e923ccdbba

If we were to add the new protocol, that would have to be a separate PR, but I think it tidies things up quite a bit.